python/test/regression/test_cast_matmul.py

```diff
@@ -57,15 +56,14 @@ def matmul_kernel(A, B, C, M, N, K,  #
     # pointers
     A = A + (ram[:, None] * stride_am + rk[None, :] * stride_ak)
     B = B + (rk[:, None] * stride_bk + rbn[None, :] * stride_bn)
-    acc = tl.zeros((BLOCK_M, BLOCK_N), dtype=dot_out_dtype)
+    acc_dtype = tl.float16 if compute_dtype == tl.float16 and C.dtype.element_ty == tl.float16 else tl.float32
+    acc = tl.zeros((BLOCK_M, BLOCK_N), dtype=acc_dtype)
     for k in range(0, tl.cdiv(K, BLOCK_K)):
         k_remaining = K - k * BLOCK_K
-        _0 = tl.zeros((1, 1), dtype=C.dtype.element_ty)
+        _0 = tl.zeros((1, 1), dtype=compute_dtype)
         a = tl.load(A, mask=rk[None, :] < k_remaining, other=_0)
         b = tl.load(B, mask=rk[:, None] < k_remaining, other=_0)
-        a = a.to(C.dtype.element_ty)
-        b = b.to(C.dtype.element_ty)
-        acc += tl.dot(a, b, out_dtype=dot_out_dtype)
+        acc += tl.dot(a.to(compute_dtype), b.to(compute_dtype), out_dtype=acc_dtype)
```

Comment:
We now tests cases with `out_dtype != compute_dtype` as these are quite common (fp16xfp16 into fp32)